lib/concurrent/executor/executor.rb

```diff
@@ -5,12 +5,75 @@
 module Concurrent
 
   module Executor
+    class AtExitAutoTermination
+      def initialize(enabled = true)
+        @mutex     = Mutex.new
+        @executors = Set.new
+        @enabled   = enabled
+        @installed = false
+      end
+
+      def add(executor)
+        @mutex.synchronize { @executors.add executor }
+      end
+
+      def remove(executor)
+        @mutex.synchronize { @executors.delete executor }
+      end
+
+      def enabled?
+        @mutex.synchronize { @enabled }
+      end
+
+      def enabled=(value)
+        @mutex.synchronize { @enabled = value }
+      end
+
+      def terminate
+        executors = @mutex.synchronize { @executors.to_a if @enabled }
+        if executors
+          executors.each(&:kill) # TODO be gentle first
+          executors.each(&:wait_for_termination)
+          executors.clear
+          true
+        else
+          false
+        end
+      end
+
+      def install
+        @installed ||= begin
+          at_exit { terminate }
+          true
+        end
+        self
+      end
+
+      def executors
+        @mutex.synchronize { @executors.to_a }
+      end
+    end
+
+    AT_EXIT_AUTO_TERMINATION = AtExitAutoTermination.new.install
```

Comment:
The current implementation is fine and I'm comfortable with us merging to master.

I don't like creeping complexity. The original `at_exit` handlers were a part of the Java thread pools since the initial implementation and no one cared. No one even noticed until we discovered a bug with autotest. In the end that bug had nothing to do with the `at_exit` handlers--it was a bug in Actor caused by an incorrect use of Delay that was, in turn, caused by a change in the default behavior of Delay. We've solved those problems. Had that bug not existed we may never have changed the `at_exit` handlers at all. Instead, we added an increasing level of complexity, added more global objects, and tightly coupled thread pools to a global registry--all so we can have the same behavior we had *before* we changed anything. I appreciate that the old implementation created a small memory leak, but I can't imagine any circumstance under which that wouldn't be trivial. No application should create more than a few thread pools. I'm not convinced that the added complexity is really warranted. That being said, I realize that my opinion is only one of many. If others are more comfortable with a global registry for shutting down thread pools then I'm fine with it, too (at least one or two others favored this approach). My overarching concern has always been the unfortunate behavior on JRuby where applications will fail to exit if the thread pools aren't properly shut down. Since we create global thread pools in the background I feel strongly that we have a responsibility to manage those thread pools for our users. This solution solves that problem so I'm good.